sopel/tools/identifiers.py

```diff
@@ -0,0 +1,266 @@
+others. To ensure that two channels or two users are the same, their
+identifiers must be processed to be compared properly. This process depends on
+which RFC and how that RFC is implemented by the server: IRC being an old
+protocol, different RFCs have differents version of that process:
+
+* :rfc:`1549#section-2.2`: ASCII characters, and ``[]\\`` are mapped to ``{}|``
+* :rfc:`2812#section-2.2`: same as in the previous RFC, adding ``~`` mapped to
+  ``^``
+
+Then when ISUPPORT was added, the `CASEMAPPING parameter`__ was defined so the
+server can say which process to apply:
+
+* ``ascii``: only ``[A-Z]`` must be mapped to ``[a-z]`` (implemented by
+  :func:`ascii_lower`)
+* ``rfc1459``: follow :rfc:`2812`; because of how it was implemented in most
+  server (implemented by :func:`rfc1459_lower`)
+* A strict version of :rfc:`1459` also exist but it is not recommended
+  (implemented by :func:`rfc1459_strict_lower`)
+
+As a result, the :class:`Identifier` class requires a casemapping function,
+which should be provided by the :class:`bot<sopel.bot.Sopel>`.
+
+.. seealso::
+
+    The bot's :class:`make_identifier<sopel.bot.Sopel.make_identifier>` method
+    should be used to instantiate an :class:`Identifier` to honor the
+    ``CASEMAPPING`` parameter.
+
+.. __: https://modern.ircdocs.horse/index.html#casemapping-parameter
+"""
+from __future__ import generator_stop
+
+import string
+from typing import Callable
+
+Casemapping = Callable[[str], str]
+
+ASCII_TABLE = str.maketrans(string.ascii_uppercase, string.ascii_lowercase)
+RFC1459_TABLE = str.maketrans(
+    string.ascii_uppercase + '[]\\~',
+    string.ascii_lowercase + '{}|^',
+)
+RFC1459_STRICT_TABLE = str.maketrans(
+    string.ascii_uppercase + '[]\\',
+    string.ascii_lowercase + '{}|',
+)
+
+
+def ascii_lower(text: str) -> str:
+    """Lower ``text`` according to the ``ascii`` value of ``CASEMAPPING``.
+
+    In that version, only ``[A-Z]`` are to be mapped to their lowercase
+    equivalent (``[a-z]``). Non-ASCII characters are kept unmodified.
+    """
+    return text.translate(ASCII_TABLE)
+
+
+def rfc1459_lower(text: str) -> str:
+    """Lower ``text`` according to :rfc:`2812`.
+
+    Similar to :func:`rfc1459_strict_lower`, but also maps ``~`` to ``^``, as
+    per :rfc:`2812#section-2.2`:
+
+        Because of IRC's Scandinavian origin, the characters ``{}|^`` are
+        considered to be the lower case equivalents of the characters
+        ``[]\\~``, respectively.
+
+    .. note::
+
+        This is an implementation of the `CASEMAPPING parameter`__ for the
+        value ``rfc1459``, which doesn't use :rfc:`1459` but its updated version
+        :rfc:`2812`.
+
+    .. __: https://modern.ircdocs.horse/index.html#casemapping-parameter
+    """
+    return text.translate(RFC1459_TABLE)
+
+
+def rfc1459_strict_lower(text: str) -> str:
+    """Lower ``text`` according to :rfc:`1459` (strict version).
+
+    As per :rfc:`1459#section-2.2`:
```

Comment:
Anywhere a specific section is linked like this, it would be useful to override the display text, e.g. "RFC 1459 § 2.2", rather than having the `#section-x.y` fragment identifier included in the final docs.

I am, in fact, about to research how annoying it would be to define our own custom role that takes an RFC number _and_ section number, and formats it appropriately without having to muck with manual overrides every time.

Reply:
Oh oh oh, good luck with that. I know what it would take, and I'm not super into it. 😁

Reply:
Just fix the rendering for this PR by overriding the display text, and let me worry about scratching the Sphinx-extension itch if I start feeling it really strongly. 😉

In this case, the below would implement the example I wrote above:

```suggestion
    As per :rfc:`RFC 1459 § 2.2<1459#section-2.2>`:
```

Which renders a link to `https://tools.ietf.org/html/rfc1459.html#section-2.2` like this (on Alabaster):

![image](https://user-images.githubusercontent.com/164140/149685303-75623e2e-0251-46dd-8c4e-39e64a4239d1.png)

(Link rendering shouldn't differ between Alabaster and Furo, so I didn't bother changing branches.)